File: scripts/verify-liblib-batch92.mjs
import assert from "node:assert/strict";
import fs from "node:fs";
import {
  DIRECTOR_LOCAL_RESOURCE_MAX_BYTES,
  addDirectorLocalResource,
  beginDirectorLocalResourceLoad,
  createDirectorLocalResourceDescriptor,
  createDirectorLocalResourceMap,
  markDirectorLocalResourceReleased,
  releaseDirectorLocalResourceLease,
  retainDirectorLocalResource,
  retryDirectorLocalResource,
  settleDirectorLocalResource,
} from "../src/lib/directorLocalResourceLifecycle.ts";

const root = new URL("..", import.meta.url);
const lifecycleSource = fs.readFileSync(
  new URL("./src/lib/directorLocalResourceLifecycle.ts", root),
  "utf8",
);
const materializerSource = fs.readFileSync(
  new URL("./src/lib/directorLocalModelMaterializer.ts", root),
  "utf8",
);

const dataUrl = "data:text/plain;base64,dGVzdA==";
const validItem = {
  id: "batch92-resource",
  fileName: "batch92.obj",
  dataUrl,
  mimeType: "text/plain",
  sizeBytes: 4,
  lastModified: 1,
  categoryId: "my-models",
  name: "Batch 92",
  visual: "box",
  color: "#8899aa",
};
const ownerA = {
  ownerKey: "libtv:batch92-canvas:batch92-source",
  projectId: "batch92-project",
  sessionId: "batch92-session-a",
  generation: 1,
};
const ownerB = {
  ...ownerA,
  sessionId: "batch92-session-b",
  generation: 2,
};

const descriptor = createDirectorLocalResourceDescriptor(validItem);
assert.ok(descriptor);
assert.equal(descriptor.sizeBytes, 4);
assert.equal(
  createDirectorLocalResourceDescriptor({
    ...validItem,
    sizeBytes: 5,
  }),
  null,
);
assert.equal(
  createDirectorLocalResourceDescriptor({
    ...validItem,
    dataUrl: "data:text/plain;base64,%%%%",
  }),
  null,
);
assert.equal(
  createDirectorLocalResourceDescriptor({
    ...validItem,
    sizeBytes: DIRECTOR_LOCAL_RESOURCE_MAX_BYTES + 1,
  }),
  null,
);
assert.equal(
  createDirectorLocalResourceDescriptor({
    ...validItem,
    mimeType: "",
  }),
  null,
);
assert.equal(
  createDirectorLocalResourceDescriptor({
    ...validItem,
    lastModified: Number.NaN,
  }),
  null,
);

let resources = createDirectorLocalResourceMap([validItem]);
const first = beginDirectorLocalResourceLoad(
  resources,
  validItem.id,
  "request-a",
  ownerA,
);
assert.equal(first.accepted, true);
resources = { ...resources, [validItem.id]: first.state };
resources = retainDirectorLocalResource(resources, validItem.id, {
  leaseId: "lease-a",
  owner: ownerA,
});
assert.equal(resources[validItem.id]?.leaseCount, 1);
assert.equal(resources[validItem.id]?.activeRequestOwner?.sessionId, ownerA.sessionId);

const wrongOwnerSettle = settleDirectorLocalResource(resources, {
  resourceId: validItem.id,
  requestId: "request-a",
  owner: ownerB,
  status: "ready",
});
assert.equal(wrongOwnerSettle.accepted, false);
assert.equal(wrongOwnerSettle.state?.status, "loading");

const wrongOwnerCancel = settleDirectorLocalResource(resources, {
  resourceId: validItem.id,
  requestId: "request-a",
  owner: ownerB,
  status: "canceled",
  error: "ABORTED",
  errorMessage: "wrong owner",
});
assert.equal(wrongOwnerCancel.accepted, false);

const wrongOwnerRelease = releaseDirectorLocalResourceLease(
  resources,
  validItem.id,
  "lease-a",
  ownerB,
);
assert.equal(wrongOwnerRelease[validItem.id]?.leaseCount, 1);

const ready = settleDirectorLocalResource(resources, {
  resourceId: validItem.id,
  requestId: "request-a",
  owner: ownerA,
  status: "ready",
});
assert.equal(ready.accepted, true);
assert.equal(ready.state?.error, null);
resources = { ...resources, [validItem.id]: ready.state };

const duplicateLease = retainDirectorLocalResource(resources, validItem.id, {
  leaseId: "lease-a",
  owner: ownerA,
});
assert.equal(duplicateLease[validItem.id]?.leaseCount, 1);

resources = markDirectorLocalResourceReleased(resources, validItem.id);
assert.equal(resources[validItem.id]?.status, "ready");
assert.equal(resources[validItem.id]?.releaseRequested, true);
const blockedRetry = retryDirectorLocalResource(resources, validItem.id);
assert.equal(blockedRetry[validItem.id]?.retryNonce, 0);
resources = releaseDirectorLocalResourceLease(
  resources,
  validItem.id,
  "lease-a",
  ownerA,
);
assert.equal(resources[validItem.id]?.status, "released");
assert.equal(resources[validItem.id]?.leaseCount, 0);
assert.equal(resources[validItem.id]?.releaseRequested, true);

resources = addDirectorLocalResource(resources, descriptor);
assert.equal(resources[validItem.id]?.status, "idle");
assert.equal(resources[validItem.id]?.releaseRequested, false);
assert.equal(resources[validItem.id]?.leaseCount, 0);

const second = beginDirectorLocalResourceLoad(
  resources,
  validItem.id,
  "request-b",
  ownerA,
);
assert.equal(second.accepted, true);
resources = { ...resources, [validItem.id]: second.state };
const invalidReady = settleDirectorLocalResource(resources, {
  resourceId: validItem.id,
  requestId: "request-b",
  owner: ownerA,
  status: "ready",
  error: "PARSE_FAILED",
});
assert.equal(invalidReady.accepted, false);
assert.equal(invalidReady.state?.status, "loading");
const failed = settleDirectorLocalResource(resources, {
  resourceId: validItem.id,
  requestId: "request-b",
  owner: ownerA,
  status: "failed",
  error: "PARSE_FAILED",
  errorMessage: "fixture parse failure",
});
assert.equal(failed.accepted, true);
resources = { ...resources, [validItem.id]: failed.state };
resources = retainDirectorLocalResource(resources, validItem.id, {
  leaseId: "lease-b",
  owner: ownerA,
});
resources = releaseDirectorLocalResourceLease(
  resources,
  validItem.id,
  "lease-b",
  ownerA,
);
resources = retryDirectorLocalResource(resources, validItem.id);
assert.equal(resources[validItem.id]?.status, "idle");

assert.match(lifecycleSource, /releaseRequested: boolean/);
assert.match(lifecycleSource, /sameLeaseOwner/);
assert.match(lifecycleSource, /DIRECTOR_LOCAL_RESOURCE_MAX_BYTES/);
assert.match(materializerSource, /bytes\.byteLength > DIRECTOR_LOCAL_RESOURCE_MAX_BYTES/);

console.log(
  JSON.stringify({
    batch: 92,
    status: "PASS",
    cases: {
      strictDescriptorAndByteBudget: true,
      ownerScopedRequestAndLease: true,
      wrongOwnerZeroMutation: true,
      terminalInvariant: true,
      deferredRelease: true,
      finalLeaseRelease: true,
      releasedResourceReactivation: true,
      retryAfterFailure: true,
      materializerBudgetGuard: true,
    },
  }),
);

File: scripts/verify-liblib-batch92.py
#!/usr/bin/env python3

"""Verify Batch 92 Director local-resource owner and release boundaries."""

from __future__ import annotations

import base64
import json
import os
from pathlib import Path
from typing import Any

from playwright.sync_api import Page, sync_playwright


ROOT = Path(__file__).resolve().parents[1]
BASE_URL = os.environ.get("LIBLIB_BASE_URL", "http://localhost:4317")
AUDIT_PATH = (
    ROOT
    / "docs"
    / "research"
    / "liblib-canvas-batch92-2026-08-29"
    / "runtime-audit.json"
)
LOCAL_MODEL_KEY = "liblib-tv-director-local-model-library-v1"


def attach_errors(page: Page) -> tuple[list[str], list[str]]:
    errors: list[str] = []
    phase = ["startup"]
    page.on(
        "console",
        lambda message: errors.append(
            f"{phase[0]}:console:{message.type}:{message.text}"
        )
        if message.type == "error"
        else None,
    )
    page.on(
        "pageerror",
        lambda error: errors.append(
            f"{phase[0]}:pageerror:{error}\n{error.stack or 'stack unavailable'}"
        ),
    )
    page.on(
        "requestfailed",
        lambda request: errors.append(
            f"{phase[0]}:requestfailed:{request.method}:{request.url}:"
            f"{request.failure}"
        ),
    )
    return errors, phase


def wait_for_app(page: Page) -> None:
    page.wait_for_function(
        "() => Boolean(window.__libtv_store && window.__director_store)"
    )
    page.wait_for_selector("[data-libtv-react-flow-host]")


def open_director(page: Page) -> dict[str, str]:
    owner = page.evaluate(
        """() => {
          const canvasState = window.__libtv_store.getState();
          const canvas = canvasState.canvases.find((item) =>
            item.nodes.some((node) => node.type === "script-execution")
          );
          if (!canvas) throw new Error("Director canvas fixture is missing");
          canvasState.setActiveCanvas(canvas.id);
          const source = canvas.nodes.find(
            (node) => node.type === "script-execution"
          );
          if (!source) throw new Error("Director source fixture is missing");
          return { canvasId: canvas.id, sourceNodeId: source.id };
        }"""
    )
    page.evaluate(
        """(owner) => {
          window.__libtv_ui_store.getState().openDirectorDesk(
            owner.sourceNodeId,
            owner.canvasId
          );
        }""",
        owner,
    )
    page.locator("[data-director-workspace]").wait_for(state="visible")
    page.wait_for_function(
        """(owner) => {
          const state = window.__director_store.getState();
          return state.projectOwner?.canvasId === owner.canvasId &&
            state.projectOwner?.sourceNodeId === owner.sourceNodeId &&
            state.projectId !== null &&
            state.sessionId !== null &&
            state.generation !== null &&
            state.projectLifecycle === "ACTIVE";
        }""",
        arg=owner,
    )
    page.locator('canvas[data-director-webgl-canvas="true"]').wait_for(
        state="visible"
    )
    return owner


def read_resource(page: Page, resource_id: str) -> dict[str, Any]:
    return page.evaluate(
        """(resourceId) => {
          const resource =
            window.__director_store.getState().localModelResources[resourceId];
          if (!resource) throw new Error(`resource missing: ${resourceId}`);
          return resource;
        }""",
        resource_id,
    )


def main() -> None:
    obj = "\n".join(
        [
            "o batch92-triangle",
            "v -0.6 0 0",
            "v 0.6 0 0",
            "v 0 1.2 0",
            "f 1 2 3",
        ]
    ).encode("utf-8")
    data_url = "data:text/plain;base64," + base64.b64encode(obj).decode("ascii")

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch()
        context = browser.new_context(viewport={"width": 1440, "height": 900})
        page = context.new_page()
        errors, phase = attach_errors(page)
        page.goto(BASE_URL, wait_until="domcontentloaded")
        wait_for_app(page)

        phase[0] = "reset"
        page.evaluate(
            """(modelKey) => {
              for (const key of Object.keys(localStorage)) {
                if (key.startsWith("liblib-tv-director-project-v1:")) {
                  localStorage.removeItem(key);
                }
              }
              localStorage.removeItem(modelKey);
            }""",
            LOCAL_MODEL_KEY,
        )
        page.reload(wait_until="domcontentloaded")
        wait_for_app(page)
        open_director(page)

        item = {
            "id": "batch92-release-resource",
            "categoryId": "my-models",
            "name": "Batch 92 release",
            "fileName": "batch92-release.obj",
            "dataUrl": data_url,
            "mimeType": "text/plain",
            "sizeBytes": len(obj),
            "lastModified": 1,
            "visual": "box",
            "color": "#7db5d8",
        }

        phase[0] = "materialize-and-lease"
        object_id = page.evaluate(
            """(item) => {
              const state = window.__director_store.getState();
              state.addLocalModelLibraryItem(item);
              return state.addModelLibraryObject(item);
            }""",
            item,
        )
        page.wait_for_function(
            """(resourceId) =>
              window.__director_store.getState().localModelResources[resourceId]?.status === "ready" &&
              window.__director_store.getState().localModelResources[resourceId]?.leaseCount === 1 &&
              window.__director_store.getState().localModelResources[resourceId]?.leases.length === 1""",
            arg=item["id"],
        )
        ready = read_resource(page, item["id"])
        assert ready["activeRequestId"] is None, ready
        assert ready["activeRequestOwner"] is None, ready
        assert ready["leases"][0]["owner"]["projectId"] == page.evaluate(
            "() => window.__director_store.getState().projectId"
        )
        assert object_id in page.evaluate(
            "() => window.__director_store.getState().authoredObjects.map((object) => object.id)"
        )

        phase[0] = "deferred-release"
        page.evaluate(
            """(resourceId) => {
              const state = window.__director_store.getState();
              state.releaseLocalModelResource(resourceId);
            }""",
            item["id"],
        )
        deferred = read_resource(page, item["id"])
        assert deferred["status"] == "ready", deferred
        assert deferred["releaseRequested"] is True, deferred
        assert deferred["leaseCount"] == 1, deferred

        phase[0] = "delete-and-final-release"
        result = page.evaluate(
            """(resourceId) =>
              window.__director_store.getState().removeLocalModelLibraryItem(
                resourceId,
                "CASCADE"
              )""",
            item["id"],
        )
        assert result["disposition"] == "COMMITTED", result
        page.wait_for_function(
            """(resourceId) => {
              const state = window.__director_store.getState();
              const resource = state.localModelResources[resourceId];
              return !state.localModelLibrary.some((item) => item.id === resourceId) &&
                resource?.status === "released" &&
                resource.leaseCount === 0 &&
                resource.releaseRequested === true;
            }""",
            arg=item["id"],
        )
        released = read_resource(page, item["id"])
        assert released["status"] == "released", released
        assert released["leaseCount"] == 0, released
        assert released["leases"] == [], released
        assert object_id not in page.evaluate(
            "() => window.__director_store.getState().authoredObjects.map((object) => object.id)"
        )

        phase[0] = "close"
        page.evaluate(
            """() => {
              const state = window.__director_store.getState();
              if (state.projectOwner) state.closeSession(state.projectOwner);
              window.__libtv_ui_store.getState().closeDirectorDesk();
            }"""
        )
        page.locator("[data-director-workspace]").wait_for(state="hidden")

        audit = {
            "batch": 92,
            "status": "SCRIPT_RECORDED_PASS",
            "baseUrl": BASE_URL,
            "contract": {
                "scope": "clone-owned Director local-resource owner and release lifecycle",
                "sourceExact": False,
                "remoteSync": False,
                "portableBytes": False,
            },
            "desktop": {
                "readyCarriesLease": True,
                "deferredReleasePreservesLease": True,
                "deleteRemovesReference": True,
                "finalLeaseReleaseReachesReleased": True,
                "releasedStateHasNoLeases": True,
            },
            "diagnostics": {
                "consoleErrors": len(
                    [item for item in errors if ":console:" in item]
                ),
                "pageErrors": len(
                    [item for item in errors if ":pageerror:" in item]
                ),
                "requestFailures": len(
                    [item for item in errors if ":requestfailed:" in item]
                ),
                "details": errors,
            },
        }
        AUDIT_PATH.write_text(
            json.dumps(audit, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
        assert not errors, errors
        print(json.dumps(audit, ensure_ascii=False, separators=(",", ":")))
        context.close()
        browser.close()


if __name__ == "__main__":
    main()
